gateway/src/endpoints/batch_inference.rs

```diff
@@ -0,0 +1,886 @@
+    fields(
+        function_name = %params.function_name,
+        variant_name = ?params.variant_name,
+    )
+)]
+#[debug_handler(state = AppStateData)]
+pub async fn prepare_batch_inference_handler(
+    State(AppStateData {
+        config,
+        http_client,
+        clickhouse_connection_info,
+    }): AppState,
+    StructuredJson(params): StructuredJson<Params>,
+) -> Result<Response<Body>, Error> {
+    // Get the function config or return an error if it doesn't exist
+    let function = config.get_function(&params.function_name)?;
+    let num_inferences = params.inputs.len();
+    if num_inferences == 0 {
+        return Err(ErrorDetails::InvalidRequest {
+            message: "No inputs provided".to_string(),
+        }
+        .into());
+    }
+    let batch_dynamic_tool_params: Vec<DynamicToolParams> =
+        BatchDynamicToolParamsWithSize(params.dynamic_tool_params, num_inferences).try_into()?;
+    let batch_dynamic_output_schemas: Vec<Option<DynamicJSONSchema>> =
+        BatchOutputSchemasWithSize(params.output_schemas, num_inferences).try_into()?;
+
+    let tool_configs = batch_dynamic_tool_params
+        .into_iter()
+        .map(|dynamic_tool_params| function.prepare_tool_config(dynamic_tool_params, &config.tools))
+        .collect::<Result<Vec<_>, _>>()?;
+    // Collect the function variant names as a Vec<&str>
+    let mut candidate_variant_names: Vec<&str> =
+        function.variants().keys().map(AsRef::as_ref).collect();
+
+    // If the function has no variants, return an error
+    if candidate_variant_names.is_empty() {
+        return Err(ErrorDetails::InvalidFunctionVariants {
+            message: format!("Function `{}` has no variants", params.function_name),
+        }
+        .into());
+    }
+
+    // Validate the input
+    params
+        .inputs
+        .iter()
+        .enumerate()
+        .try_for_each(|(i, input)| {
+            function.validate_input(input).map_err(|e| {
+                Error::new(ErrorDetails::BatchInputValidation {
+                    index: i,
+                    message: e.to_string(),
+                })
+            })
+        })?;
+
+    // If a variant is pinned, only that variant should be attempted
+    if let Some(ref variant_name) = params.variant_name {
+        candidate_variant_names.retain(|k| k == variant_name);
+
+        // If the pinned variant doesn't exist, return an error
+        if candidate_variant_names.is_empty() {
+            return Err(ErrorDetails::UnknownVariant {
+                name: variant_name.to_string(),
+            }
+            .into());
+        }
+    }
+
+    // Retrieve or generate the episode IDs and validate them (in the impl)
+    let episode_ids: BatchEpisodeIds =
+        BatchEpisodeIdsWithSize(params.episode_ids, num_inferences).try_into()?;
+
+    // Increment the request count
+    counter!(
+        "request_count",
+        "endpoint" => "post_batch_inference",
```

Comment:
`endpoint` should be the actual endpoint (`batch_inference`)

We can optionally further parametrize with `method` = `POST`. If we do that, we should do that for every endpoint.